requests/sessions.py

```diff
@@ -147,8 +148,12 @@ def resolve_redirects(self, resp, req, stream=False, timeout=None,
                 parsed_rurl = urlparse(resp.url)
                 url = '%s:%s' % (to_native_string(parsed_rurl.scheme), url)
 
-            # The scheme should be lower case...
+            # Normalize url case and attach previous fragment if needed (RFC 7231 7.1.2)
             parsed = urlparse(url)
+            if parsed.fragment == '' and previous_fragment:
+                parsed = urlparse('{0}#{1}'.format(url, previous_fragment))
```

Comment:
Can we do `parsed._replace(fragment='previous_fragment')`?

Reply:
My concern is that `url` might end with `#` which should be equivalent to `parsed.fragment == ''`. What do we do there?

Reply:
Yep, this is definitely a gap. It looks like `geturl()` will also strip empty fragments from redirect urls which probably isn’t the behavior we want.

I’m a little leery of using private methods, but given that `_replace` has been stable for a decade it’s probably safe. I’m wondering given the trailing hash issue if it may be easier to append the fragment option (`’’`, `#`, `#fragment`) after we call `geturl`. I’ll try to come up with something tonight.

Reply:
`_replace` is actually part of `namedtuple` which `ParseResult` inherits from. If we wanted to, we could switch wholesale to github.com/hyper/rfc3986 which has a proper public API for this but that is definitely more work than this.